synfig-studio/src/gui/dialogs/dialog_pluginmanager.cpp

```diff
@@ -0,0 +1,252 @@
+{
+    // Show the dialog and wait for a user response
+    int result = plugin_file_dialog.run();
+    switch (result) {
+        case (Gtk::RESPONSE_OK):
+        { 
+            std::vector<std::string> files;
+            etl::handle<synfig::FileSystemNative> native_fs = FileSystemNative::instance();
+            etl::handle<synfig::FileContainerZip> zip_fs = new FileContainerZip();
+            std::string zip_filename = plugin_file_dialog.get_filename();
+            filesystem::Path path_to_user_plugins = synfigapp::Main::get_user_app_directory() / filesystem::Path("plugins");
+            if (!zip_fs->open(plugin_file_dialog.get_filename())) {
+                synfig::error("Failed to open zip file: " + zip_filename);
+                return;
+            }
+            zip_fs->directory_scan("", files );
+            // check if the archive has a plugin.xml file at root directory
+            enum PluginZipStatus
+            {
+                PLUGIN_ZIP_NO_PLUGIN_XML = 1 << 1,
+                PLUGIN_ZIP_AT_ROOT_DIR = 1 << 2,
+                PLUGIN_ZIP_AT_CHILD_DIR = 1 << 3
+            } pluginStatus = PluginZipStatus::PLUGIN_ZIP_NO_PLUGIN_XML;
+            
+            for(auto& file : files) {
+                if(file == "plugin.xml") {
+                    pluginStatus = PluginZipStatus::PLUGIN_ZIP_AT_ROOT_DIR;
+                    break;
+                }
+            }
+            std::string output_path ;
+            if(pluginStatus == PluginZipStatus::PLUGIN_ZIP_AT_ROOT_DIR) {
+                output_path = ((path_to_user_plugins / filesystem::Path(zip_filename).stem()).add_suffix("/")).u8string();
+                if(native_fs->directory_create(output_path))
+                for(auto &file : files)
+                    zip_fs->copy_recursive(zip_fs, file, native_fs, output_path + file);
+                App::plugin_manager.load_plugin(output_path + "plugin.xml", output_path, true);
+            }
+            else {
+                for(auto& file : files)
+                    if(zip_fs->is_directory(file)) {
+                        std::vector<std::string> child_files;
+                        zip_fs->directory_scan(file, child_files);
+                        for (auto& file : child_files)
+                            if(file == "plugin.xml") {
+                                pluginStatus = PluginZipStatus::PLUGIN_ZIP_AT_CHILD_DIR;
+                                break;
+                            }
+                    }
+                if(pluginStatus == PluginZipStatus::PLUGIN_ZIP_AT_CHILD_DIR) {
+                    output_path = path_to_user_plugins.add_suffix("/").u8string();
+                    for(auto &file : files)
+                        zip_fs->copy_recursive(zip_fs, file, native_fs, output_path + file);
+                }
+                App::plugin_manager.load_plugin(output_path + files[0] + '/' + "plugin.xml", output_path +  files[0] , true);
```

Comment:
You could create a method or function just for searching the plugin.xml file and the optional directory it could be in.
Then, you would do the extraction.


I think the installed plugin folder could be named after the plugin metadata info, instead of zip filename or zip internal folder. But it can be done later.